pallets/randomness-beacon/src/lib.rs

```diff
@@ -210,66 +239,96 @@ pub mod pallet {
-			Ok(())
+		fn check_inherent(call: &Self::Call, _data: &InherentData) -> Result<(), Self::Error> {
+			match call {
+				Call::try_submit_asig { .. } => Ok(()),
+				_ => unreachable!("other calls are not inherents"),
+			}
 		}
 
 		fn is_inherent(call: &Self::Call) -> bool {
 			matches!(call, Call::try_submit_asig { .. })
 		}
 	}
 
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// A dummy `on_initialize` to return the amount of weight that `on_finalize` requires to
+		/// execute.
+		fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+			// weight of `on_finalize`
+			T::WeightInfo::on_finalize()
+		}
+
+		/// At the end of block execution, the `on_finalize` hook checks that the timestamp was
+		/// updated. Upon success, it removes the boolean value from storage. If the value resolves
+		/// to `false`, the pallet will panic.
+		///
+		/// ## Complexity
+		/// - `O(1)`
+		fn on_finalize(_n: BlockNumberFor<T>) {
+			assert!(
+				DidUpdate::<T>::take(),
+				"The aggregated siganture must be updated once in the block"
+			);
+		}
+	}
+
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
 		/// Write a set of pulses to the runtime
 		///
 		/// * `origin`: A None origin
 		/// * `asig`: An aggregated signature
+		/// * `height`: The number of sigs aggregated to construct asig
 		/// * `round`: An optional genesis round number. It can only be set if the existing genesis
 		///   round is 0.
 		#[pallet::call_index(0)]
 		#[pallet::weight(T::WeightInfo::try_submit_asig())]
 		pub fn try_submit_asig(
 			origin: OriginFor<T>,
 			asig: OpaqueSignature,
+			height: RoundNumber,
 			round: Option<RoundNumber>,
 		) -> DispatchResult {
-			// In the future, this will expected a signed payload
-			// https://github.com/ideal-lab5/idn-sdk/issues/117
 			ensure_none(origin)?;
+			ensure!(!DidUpdate::<T>::exists(), Error::<T>::SignatureAlreadyVerified,);
+
 			let config = T::BeaconConfig::get();
 			let mut genesis_round = GenesisRound::<T>::get();
 			let mut latest_round = LatestRound::<T>::get();
 
+			ensure!(height > 0, Error::<T>::ZeroHeightProvided);
+			ensure!(
+				height <= T::MaxSigsPerBlock::get() as u64,
+				Error::<T>::ExcessiveHeightProvided
+			);
+
 			if let Some(r) = round {
 				// if a round is provided and the genesis round is not set
-				frame_support::ensure!(genesis_round == 0, Error::<T>::GenesisRoundAlreadySet);
+				ensure!(genesis_round == 0, Error::<T>::GenesisRoundAlreadySet);
 				GenesisRound::<T>::set(r);
 				genesis_round = r;
 				latest_round = genesis_round;
 			} else {
 				//  if the genesis round is not set and a round is not provided
-				frame_support::ensure!(
-					GenesisRound::<T>::get() > 0,
-					Error::<T>::GenesisRoundNotSet
-				);
+				ensure!(GenesisRound::<T>::get() > 0, Error::<T>::GenesisRoundNotSet);
 			}
-
 			// aggregate old asig/apk with the new one and verify the aggregation
+			// Q: do we care about the entire linear history of message hashes?
```

Comment:
This is useful for searching for the keyword 'todo'
```suggestion
			// TODO: do we care about the entire linear history of message hashes?
```